quickwit/quickwit-control-plane/src/indexing_scheduler/scheduling/scheduling_logic.rs

```diff
@@ -0,0 +1,620 @@
+            node_available_capacity[node_id] = node_assignment.node_available_capacity(problem);
+            if num_shards_to_remove == 0 {
+                // No more work to do for this source.
+                break;
+            }
+        }
+    }
+    assert_remove_extraneous_shards_post_condition(problem, solution);
+}
+
+fn assert_remove_extraneous_shards_post_condition(
+    problem: &SchedulingProblem,
+    solution: &SchedulingSolution,
+) {
+    let mut num_shards_per_source: Vec<u32> = vec![0; problem.num_sources()];
+    for node_assignment in &solution.node_assignments {
+        for (&source, &load) in &node_assignment.num_shards_per_source {
+            num_shards_per_source[source as usize] += load;
+        }
+    }
+    for source in problem.sources() {
+        assert!(num_shards_per_source[source.source_id as usize] <= source.num_shards);
+    }
+}
+
+// -------------------------------------------------------------------------
+// Phase 2
+// Releave sources from the node that are exceeding their maximum load.
+
+fn enforce_nodes_max_load(problem: &SchedulingProblem, solution: &mut SchedulingSolution) {
+    for node_assignment in solution.node_assignments.iter_mut() {
+        let node_max_load: Load = problem.node_max_load(node_assignment.node_id);
+        enforce_node_max_load(problem, node_max_load, node_assignment);
+    }
+}
+
+fn enforce_node_max_load(
+    problem: &SchedulingProblem,
+    node_max_load: Load,
+    node_assignment: &mut NodeAssignment,
+) {
+    let total_load = node_assignment.total_load(problem);
+    if total_load <= node_max_load {
+        return;
+    }
+    let mut load_to_remove = total_load - node_max_load;
+    let mut load_sources: Vec<(Load, SourceOrd)> = node_assignment
+        .num_shards_per_source
+        .iter()
+        .map(|(source_id, num_shards)| {
+            let load_for_source = problem.source_load_per_shard(*source_id).get() * num_shards;
+            (load_for_source, *source_id)
+        })
+        .collect();
+    load_sources.sort();
+    for (load, source_id) in load_sources {
+        node_assignment.num_shards_per_source.remove(&source_id);
+        load_to_remove = load_to_remove.saturating_sub(load);
+        if load_to_remove == 0 {
+            break;
+        }
+    }
+    assert_enforce_nodes_max_load_post_condition(problem, node_assignment);
+}
+
+fn assert_enforce_nodes_max_load_post_condition(
+    problem: &SchedulingProblem,
+    node_assignment: &NodeAssignment,
+) {
+    let total_load = node_assignment.total_load(problem);
+    assert!(total_load <= problem.node_max_load(node_assignment.node_id));
+}
+
+// ----------------------------------------------------
+// Phase 3
+// Place unassigned sources.
+//
+// We use a greedy algorithm as a simple heuristic here.
+// We go through the sources in decreasing order of their load.
+//
+// We then try to place as many shards as possible in the node with the
+// highest load.
```

Comment:
From the code I read, we place as many shards as possible in the node with the highest available capacity, not the highest load.